include/eve/module/core/regular/impl/simd/riscv/if_else.hpp

```diff
@@ -0,0 +1,63 @@
+//==================================================================================================
+/*
+  EVE - Expressive Vector Engine
+  Copyright : EVE Project Contributors
+  SPDX-License-Identifier: BSL-1.0
+*/
+//==================================================================================================
+#pragma once
+
+#include <eve/concept/value.hpp>
+#include <eve/detail/function/simd/riscv/make.hpp>
+#include <eve/detail/implementation.hpp>
+
+namespace eve::detail
+{
+
+template<scalar_value T, typename N>
+EVE_FORCEINLINE wide<T, N>
+                if_else_(EVE_SUPPORTS(rvv_),
+                         logical<wide<T, N>> const                &c,
+                         wide<T, N> const                         &vt,
+                         wide<T, N> const                         &vf) noexcept
+requires rvv_abi<abi_t<T, N>>
+{
+  return __riscv_vmerge_tu(vt, vf, vt, c, N::value);
+}
+
+template<scalar_value T, typename N>
+EVE_FORCEINLINE wide<T, N>
+                if_else_(EVE_SUPPORTS(rvv_),
+                         logical<wide<T, N>> const                &c,
+                         T const                                  &vt,
+                         wide<T, N> const                         &vf) noexcept
+requires rvv_abi<abi_t<T, N>>
+{
+  if constexpr( std::is_floating_point_v<T> ) return __riscv_vfmerge(vf, vt, c, N::value);
+  else return __riscv_vmerge(vf, vt, c, N::value);
+}
+
+template<scalar_value T, typename N>
+EVE_FORCEINLINE wide<T, N>
+if_else_(EVE_SUPPORTS(rvv_), logical<wide<T, N>> const& c, T const& vt, T const& vf) noexcept
+requires rvv_abi<abi_t<T, N>>
+{
+  auto fwide = make(as<wide<T, N>> {}, vf);
+  return if_else(c, vt, fwide);
+}
```

Comment:
This should work out of the box, I'm surprised you need it.

Reply:
Otherwise I have recursion: for implementing function `to_bits` and similar I use `if_else` with constants.

Reply:
The decisions that lead to this seem incorrect.

Reply:
Make an if constexpr chain plz

Reply:
I'm not sure I understand you correctly. To be in context, here is how used target-dependent hooks on other archs in EVE:
arm neon:
* arguments `logical`, `logical`, `logical`:
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/x86/if_else.hpp#L103

arm sve:
* arguments `logical`, `wide`, `wide`.
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/arm/sve/if_else.hpp#L17

* arguments `logical`, `logical`, `logical`:
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/arm/sve/if_else.hpp#L29

ppc:
* arguments `logical`, `wide`, `wide`.
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/ppc/if_else.hpp#L18
* arguments `logical`, `logical`, `logical`:
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/ppc/if_else.hpp#L28

x86:
* arguments `logical`, `wide`, `wide`.
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/x86/if_else.hpp#L26
* arguments `logical`, `logical`, `logical`:
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/x86/if_else.hpp#L103

Do you imply that for RISC-V I should implement a different approach? Like to have the only one function:
```
template<logical_simd_value Logical, Value VT, Value VF>
requires(std::same_as<element_type_t<VT>, element_type_t<VF>> && std::same_as<element_type<typename Logical::mask_type>, element_type_t<VT>> && rvv_abi<...> && /*same cardinal, if Values are wide */)
auto if_else(..., Logical logical,, VT vt, VF vf) {
if constepxr (std::same_as<Logical, VT> && std::same_as<Logical, VF>) {...}
else if constexpr (...)
....
}
```

Also note that combining all if_else functions will require at least
quite a huge `requires` block, that may not work, as it is not working
for friend functions: compiler will not be able to distinguish
common `if_else` from riscv-specific, as both will match.

Reply:
I'm not sure I understand you correctly. To be in context, here is how used target-dependent hooks on other archs in EVE:
arm neon:
* arguments `logical`, `logical`, `logical`:
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/x86/if_else.hpp#L103

arm sve:
* arguments `logical`, `wide`, `wide`.
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/arm/sve/if_else.hpp#L17

* arguments `logical`, `logical`, `logical`:
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/arm/sve/if_else.hpp#L29

ppc:
* arguments `logical`, `wide`, `wide`.
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/ppc/if_else.hpp#L18
* arguments `logical`, `logical`, `logical`:
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/ppc/if_else.hpp#L28

x86:
* arguments `logical`, `wide`, `wide`.
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/x86/if_else.hpp#L26
* arguments `logical`, `logical`, `logical`:
https://github.com/jfalcou/eve/blob/main/include/eve/module/core/regular/impl/simd/x86/if_else.hpp#L103

Do you imply that for RISC-V I should implement a different approach? Like to have the only one function:
```
template<logical_simd_value Logical, Value VT, Value VF>
requires(std::same_as<element_type_t<VT>, element_type_t<VF>> && std::same_as<element_type<typename Logical::mask_type>, element_type_t<VT>> && rvv_abi<...> && /*same cardinal, if Values are wide */)
auto if_else(..., Logical logical,, VT vt, VF vf) {
if constepxr (std::same_as<Logical, VT> && std::same_as<Logical, VF>) {...}
else if constexpr (...)
....
}
```

Also note that combining all if_else functions will require at least
quite a huge `requires` block, that may not work, as it is not working
for friend functions: compiler will not be able to distinguish
common `if_else` from riscv-specific, as both will match.

Reply:
I see - scalar overloads.

There is a problem though. 

You have 3 of scalar overloads and you are not always calling the correct one. 
The problem comes in when the scalar type is not exactly matching T.
Then you would do a broadcast, which your code tries to avoid.

I don't see you escaping having to

```
template <logical_simd_value L, arithmetic_value T, arithmetic_value U>
auto if_else_(L, T, U);
```

If you want to overload: `wide, wide`, `wide, scalar`, `scalar, wide` - fine, but you still need to figure out casts. 

FYI: this is how you call generic version: https://github.com/jfalcou/eve/blob/89d8ab2092984c3511ecdefeed0ff9c4d96c4fdd/include/eve/module/core/compress/simd/x86/compress.hpp#L20

Reply:
> I don't see you escaping having to
template <logical_simd_value L, arithmetic_value T, arithmetic_value U>
auto if_else_(L, T, U);

Got you. Done, now for scalars we call `static_cast` to the right type, and call `if_else` again